### PR TITLE
E2e/dataset versioning milestone1

### PR DESCRIPTION
**Pull request by NT-D**
# Type of PR

- [x] Documentation changes
- [x] Code changes
- [x] Test changes
- [x] CI-CD changes
- [ ] GitHub Template changes

## Purpose

- ...
Have build e2e data pipeline to copy versioned data into Delta Lake

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Author pre-publish checklist:

- [x] Added test to prove my fix is effective or new feature works
- [ ] No PII in logs
- [ ] Made corresponding changes to the documentation

## Validation steps

- ...

## Issues Closed or Referenced

- Closes #issue_number

- References #issue_number

### CODE REVIEW

File: e2e_samples/dataset_versioning/datafactory/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Overview
+
+This folder is for Data pipeline which copy data from SQL DB to Azure Storage (Delta Lake). We are using Azure Data Factory to implement pipeline.
+
+## How to deploy it
+
+1. Provision Azure Resources by IaC (Terraform)
```
Comment (icchan) [inline e2e_samples/dataset_versioning/datafactory/README.md:7]: I think these first 2 items are not steps to deploy. Maybe they are just prerequisites since I did them before reaching this page.

File: e2e_samples/dataset_versioning/infra/README.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+**Prerequisites:**
+
+- Terraform > 0.14.7
+- Azure cli > 2.19.1
+
+**Deploy Steps:**
+
+The following deploys the Azure base infrastructure for the Temperature Events sample. Ensure you are in `e2e_samples/dataset-versioning/infra/`.
+
+1. `az login`
```
Comment (icchan) [inline e2e_samples/dataset_versioning/infra/README.md:14]: This summary is a bit confusing.
3 of them are actual commands and one is a description.

It might make more sense to for these all to just describe what the step is, the commands are below anyway.

File: e2e_samples/dataset_versioning/infra/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+Keep note of the output from `./setup.sh`. The output will be used in the next command
+
+### #3 Run terraform init with the output from step 2
+
+Use the output from the last command, to initialise the Terraform state files and store them in the Azure storage & Key Vault resources created in step 2.
+
+```bash
+# Note: This is an example, use the exact output given to you in step 2.
+az login
```
Comment (icchan) [inline e2e_samples/dataset_versioning/infra/README.md:56]: do you need to add az login again here?

File: e2e_samples/dataset_versioning/infra/README.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+terraform init -backend-config="storage_account_name=stmyterraform" -backend-config="container_name=terraform-state" -backend-config="access_key=$(az keyvault secret show --name tfstate-storage-key --vault-name kv-myterraform --query value -o tsv)" -backend-config="key=terraform.tfstate"
+```
+
+### #4 Use `terraform apply` to create the sample resources
+
+Run the following commands to deploy the sample infrastructure.
+**Note:** Pick a globally unique name and replace "your_rg_name" and "your_app_name"`
+
+```bash
+# NOTE: enter your own unique value instead of `myapp`
```
Comment (icchan) [inline e2e_samples/dataset_versioning/infra/README.md:68]: The summary above includes terraform validate, but it is not here.

I think either add it here, or remove it from the summary

File: e2e_samples/dataset_versioning/infra/README.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+terraform plan -var="rg_name=your_rg_name" -var="app_name=your_app_name" -out=/tmp/myapp
```
Comment (icchan) [inline e2e_samples/dataset_versioning/infra/README.md:69]: Not a big deal, but due to the generated naming, this line does fail if the app_name is longer than 9 characters.

File: e2e_samples/dataset_versioning/datafactory/config/README.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+- `WaterMark` table storing table name and its watermark value
+- `usp_UpdateWatermark` stored procedure to update watermark value
+
+## How to create required tables and stored procedure
+
+1. Install [Azure Data Studio](https://docs.microsoft.com/en-us/sql/azure-data-studio/download-azure-data-studio?view=sql-server-ver15)
+1. Get user id and connection string for SQL Database
+   1. Sign-in to Azure and go to provisioned resource group.
+   1. Go to Azure KeyVault
+   1. Click [Access policies] in left pane, then click [+ Add Access Policy]
```
Comment (icchan) [inline e2e_samples/dataset_versioning/datafactory/config/README.md:15]: I suggest describing what you are doing in these steps since there are a lot of steps.

Also (related to permissions)  I did not need to add an access policy and it still worked for me

File: e2e_samples/dataset_versioning/datafactory/config/README.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+   1. Click [Secret permissions] dropbox, then check [Select all] check box.
+   1. Click [None selected], then add yourself. Click [Add] to save setting.
```
Comment (icchan) [inline e2e_samples/dataset_versioning/datafactory/config/README.md:17]: I think it would be helpful to say click [None selected] next to "Select principal"

File: e2e_samples/dataset_versioning/datafactory/README.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+1. [Install the Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli)
+1. Run command to deploy ([document](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deploy-cli#deploy-local-template-or-bicep-file))
+
+```bash
```
Comment (icchan) [inline e2e_samples/dataset_versioning/datafactory/README.md:11]: Actually, we need to mention to change directories here since the arm_template.json is in ` datafactory/arm_template` but this readme is in `datafactory`

File: e2e_samples/dataset_versioning/datafactory/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+az deployment group create --name {Your deployment name. Anyname is OK} --resource-group {Your resource group name} --template-file ./arm_template.json --parameters factoryName="{Your data factory name}" KeyVault_properties_typeProperties_baseUrl="{Your key vault url}" AzureBlobFS_properties_typeProperties_serviceEndpoint="{Your blob storage url}"
+```
+
+### Parameters in deployment script
+
+|Name|Description|
+|--|--|
+|factoryName|Azure Data Factory name where you'll deploy ARM template into|
```
Comment (icchan) [inline e2e_samples/dataset_versioning/datafactory/README.md:19]: We discussed in chat but it's not very clear how to get this information. Maybe have a better description or automate if possible.

**Review by icchan — COMMENTED** [review]
These are my comments on setting up this sample.
Mostly it did work though, so I don't think there's are really blockers, just some parts are a bit hard to follow

File: e2e_samples/dataset_versioning/infra/README.md (hunk 4 above)
Comment (icchan) [reply e2e_samples/dataset_versioning/infra/README.md:68]: FYI terraform validate actually returns some warnings

│ Warning: "soft_delete_enabled": [DEPRECATED] Azure has removed support for disabling Soft Delete as of 2020-12-15, as such this field is no longer configurable and can be safely removed. This field will be removed in version 3.0 of the Azure Provider.
│
│   on ../../modules/keyvault/main.tf line 1, in resource "azurerm_key_vault" "keyvault":
│    1: resource "azurerm_key_vault" "keyvault" {
│
╵
╷
│ Warning: "extended_auditing_policy": [DEPRECATED] the `extended_auditing_policy` block has been moved to `azurerm_mssql_server_extended_auditing_policy` and `azurerm_mssql_database_extended_auditing_policy`. This block will be removed in version 3.0 of the provider.
│
│   on ../../modules/sql_server/main.tf line 45, in resource "azurerm_sql_server" "sql_server":
│   45: resource "azurerm_sql_server" "sql_server" {

File: e2e_samples/dataset_versioning/datafactory/README.md (hunk 8 above)
Comment (NT-D) [reply e2e_samples/dataset_versioning/datafactory/README.md:18]: Have updated to add path in script `./arm_tempalte/arm_template.json`

File: e2e_samples/dataset_versioning/datafactory/README.md (hunk 9 above)
Comment (NT-D) [reply e2e_samples/dataset_versioning/datafactory/README.md:26]: Have automated in this PR. A user can get deployment script when she/he provision infra successfully.
https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/305

File: e2e_samples/dataset_versioning/datafactory/README.md (hunk 1 above)
Comment (NT-D) [reply e2e_samples/dataset_versioning/datafactory/README.md]: Have updated based on this feedback. See detail in 
https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/305

File: e2e_samples/dataset_versioning/datafactory/config/README.md (hunk 6 above)
Comment (NT-D) [reply e2e_samples/dataset_versioning/datafactory/config/README.md]: Have updated. See detail in https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/305

File: e2e_samples/dataset_versioning/datafactory/config/README.md (hunk 7 above)
Comment (NT-D) [reply e2e_samples/dataset_versioning/datafactory/config/README.md]: Have updated.

File: e2e_samples/dataset_versioning/datafactory/README.md (hunk 8 above)
Comment (NT-D) [reply e2e_samples/dataset_versioning/datafactory/README.md:18]: added correct path in script.

File: e2e_samples/dataset_versioning/datafactory/README.md (hunk 9 above)
Comment (NT-D) [reply e2e_samples/dataset_versioning/datafactory/README.md:26]: Have automated in https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/305

**Review by icchan — APPROVED** [review]

**Review by akirakakar — APPROVED** [review]

**Review by devlace — CHANGES_REQUESTED** [review]

**Comment by NT-D** [comment]
Thank you @devlace for making change for this PR in PR #353

**Review by devlace — APPROVED** [review]
